easybuild/tools/run.py

```diff
@@ -75,7 +76,83 @@
 ]
 
 
-RunResult = namedtuple('RunResult', ('cmd', 'exit_code', 'output', 'stderr', 'work_dir'))
+RunShellCmdResult = namedtuple('RunShellCmdResult', ('cmd', 'exit_code', 'output', 'stderr', 'work_dir'))
+
+
+class RunShellCmdError(BaseException):
+
+    def __init__(self, cmd, exit_code, work_dir, output, stderr, caller_info, *args, **kwargs):
+        """Constructor for RunShellCmdError."""
+        self.cmd = cmd
+        self.cmd_name = cmd.split(' ')[0]
+        self.exit_code = exit_code
+        self.work_dir = work_dir
+        self.output = output
+        self.stderr = stderr
+        self.caller_info = caller_info
+
+        msg = f"Shell command '{self.cmd_name}' failed!"
+        super(RunShellCmdError, self).__init__(msg, *args, **kwargs)
+
+
+def print_run_shell_cmd_error(err):
+    """
+    Report failed shell command using provided RunShellCmdError instance
+    """
+
+    def pad_4_spaces(msg):
+        return ' ' * 4 + msg
+
+    cmd_name = err.cmd.split(' ')[0]
+    error_info = [
+        '',
+        "ERROR: Shell command failed!",
+        pad_4_spaces(f"full command              ->  {err.cmd}"),
+        pad_4_spaces(f"exit code                 ->  {err.exit_code}"),
+        pad_4_spaces(f"working directory         ->  {err.work_dir}"),
+    ]
+
+    tmpdir = tempfile.mkdtemp(prefix='shell-cmd-error-')
+    output_fp = os.path.join(tmpdir, f"{cmd_name}.out")
+    with open(output_fp, 'w') as fp:
+        fp.write(err.output or '')
```

Comment:
Well, no, since a step could be running a whole bunch of commands, and also emit way more log messages.

I think it's useful to have a log file that only includes the output of the shell command that failed, so you can start from the top and work your way down the command output without first having to ignore a bunch of other log messages that you don't care much about.

We could also considering logging the full environment in which the command was started in this command log (maybe only if `--debug` is used), I would definitely not do that in the step log...

Reply:
My bad, you're right, I thought you were referring to a log for the full step, but `run_shell_cmd` indeed also logs to `easybuild-run*log` (when `trace` is enabled), so we're logging the exact same thing twice.